feat: store star handlers

vasco-santos opened the pull request:
This PR adds the remaining `store/*` handlers: list + remove.

As follow up, created issue https://github.com/web3-storage/upload-api/issues/16 to handle the cursor moving around - this needs to be updated in spec, [access](https://github.com/web3-storage/w3protocol/blob/main/packages/access/src/capabilities/store.js#L68) and upload-api

Closes #11

alanshaw commented on api/test/service/store.test.js:
```diff
@@ -118,9 +118,161 @@ test('store add returns done if already uploaded', async (t) => {
+  const data = new Uint8Array([11, 22, 34, 44, 55])
+  const link = await CAR.codec.link(data)
+
+  const request = await CAR.encode([
```
What is the reason for using CAR encoder here? You lose the type validation from the capability if you manually create the invocation like this and also have to decode the response. For example, why not do this instead:

```js
import { connect } from '@ucanto/client'
import { CAR, CBOR } from '@ucanto/transport'
import * as StoreCapabilities from '@web3-storage/access/capabilities/store'

const conn = connect({
  id: t.context.serviceDid,
  encoder: CAR,
  decoder: CBOR,
  channel: server
})

const storeRemove = await StoreCapabilities.remove
    .invoke({
      issuer: alice,
      audience: parse(t.context.serviceDid.did()),
      capabilities: [{
        can: 'store/remove',
        with: account,
        nb: { link },
      }],
      proofs: [],
    })
    .execute(conn)
```

vasco-santos replied:
I know, the blocker for that is how to spin the SST resources in that way of doing tests. I am expecting to use client in integration tests (which performs capability invocation like you are suggesting).

However, for unit tests kind of interaction, we need to interact directly with the server instance and call request after creating it.

olizilla replied:
@vasco-santos i think Alan is saying, you can pass the uncanto server **object** directly to a uncanto client and then use the client to test different scenarios. No http requests occur, not http api gateway needed. The client object uses the server object directly.

vasco-santos replied:
In that scenario, we get limited by the things we can test on having the client to support them (example: list with pagination, ...)

I think we should use the client, but in integration tests, not unit

alanshaw replied:
If you want to test capabilities that are not in the access client _yet_ then you can always create them here in the same way as the access client creates them. You also get the benefit of being able to copy/pasta them to the client after you've defined them here and tested them.

vasco-santos replied:
Let's sync on this tomorrow and decide

olizilla replied:
i make it so in #23 and i like it

alanshaw commented on api/tables/store.js:
```diff
@@ -67,6 +70,61 @@ export function createStoreTable (region, tableName, options = {}) {
+        TableName: tableName,
+        Key: marshall({
+          uploaderDID: uploaderDID.toString(),
+          payloadCID: payloadCID.toString(),
```
Already strings according to the method params?

vasco-santos replied:
good catch, I moved things around and forgot these

alanshaw commented:
😉😉😉 "store star" sounds cooler than "store asterisk"

vasco-santos commented on api/test/service/store.test.js:
```diff
@@ -72,9 +72,9 @@ test('store add returns signed url for uploading', async (t) => {
-  t.truthy(item.uploaderDID)
+  t.truthy(item?.uploadedAt)
+  t.truthy(item?.proof)
+  t.truthy(item?.uploaderDID)
```
Made type check and added size as well

vasco-santos commented on api/tables/store.js:
```diff
@@ -67,6 +70,61 @@ export function createStoreTable (region, tableName, options = {}) {
+     * @param {string} uploaderDID
+     * @param {string} payloadCID
+     */
+     remove: async (uploaderDID, payloadCID) => {
```
oh no, CI is happy with it :(

vasco-santos commented on api/service/store/remove.js:
```diff
@@ -0,0 +1,32 @@
+      const { link } = capability.nb
+
+      if (!link) {
+        return new Server.MalformedCapability(
```
Oh ok, I did not know this. I was following old API code for it. That makes sense, I am removing

vasco-santos commented:
> 😉😉😉 "store star" sounds cooler than "store asterisk"

I like asterisk more actually, but renamed for consistency

olizilla commented on api/test/service/store.test.js:
```diff
@@ -72,9 +73,11 @@ test('store add returns signed url for uploading', async (t) => {
-  t.truthy(item.uploaderDID)
+  t.is(typeof item?.uploadedAt, 'string')
+  t.is(typeof item?.proof, 'string')
+  t.is(typeof item?.uploaderDID, 'string')
+  t.truthy(DID.parse(item?.uploaderDID))
+  t.is(typeof item?.size, 'number')
```
maybe check that it's the DID and size you expect it to be here?

olizilla commented on api/test/service/store.test.js:
```diff
@@ -117,10 +120,163 @@ test('store add returns done if already uploaded', async (t) => {
-  t.truthy(item.uploaderDID)
+  t.is(typeof item?.uploadedAt, 'string')
+  t.is(typeof item?.proof, 'string')
+  t.is(typeof item?.uploaderDID, 'string')
+  t.truthy(DID.parse(item?.uploaderDID))
+  t.is(typeof item?.size, 'number')
```
check DID is the one you expect

olizilla reviewed (approved):
👍

vasco-santos commented on api/test/service/store.test.js:
```diff
@@ -117,10 +120,163 @@ test('store add returns done if already uploaded', async (t) => {
+  t.is(typeof item?.size, 'number')
+})
+
+test('store remove does not fail for non existent link', async (t) => {
```
Discussion time in https://github.com/web3-storage/upload-api/issues/21

vasco-santos commented:
Addressed Oli's nits and merging given it is approved and also addressed all Alan's comments. 

@alanshaw we can address [this](https://github.com/web3-storage/upload-api/pull/14#discussion_r1024156583) differently if you have more feedback in follow up :)

olizilla commented on api/test/service/store.test.js:
```diff
@@ -72,9 +72,13 @@ test('store add returns signed url for uploading', async (t) => {
+  t.is(typeof item?.uploadedAt, 'string')
+  t.is(typeof item?.proof, 'string')
+  t.is(typeof item?.uploaderDID, 'string')
+  t.is(item?.uploaderDID, account)
```
account is now a bit of a trigger word for me. I see `account` and i mentally swap it with `space` and that trips me up here. Can we do
```suggestion
  t.is(item?.uploaderDID, alice.did())
```
and drop the word `account` from the tests pls?
